File: langchain-typescript/src/09-multi-agent/graph.ts
import { StateGraph, END } from "@langchain/langgraph";
import { createModelClient } from "../clients/model";
import { createSearchTool } from "../clients/tavily";
import { SupervisorAgent } from "./agents/supervisor-agent";
import { ResearcherAgent } from "./agents/researcher-agent";
import { CoderAgent } from "./agents/coder-agent";
import { ReviewerAgent } from "./agents/reviewer-agent";

interface AgentState {
  messages: Array<{ role: string; content: string }>;
  task: string;
  researchReport?: string;
  codeContent?: string;
  reviewReport?: string;
  taskType?: string;
  nextNode?: string;
}

const llm = createModelClient();
const searchTool = createSearchTool();

const supervisor = new SupervisorAgent(llm);

supervisor.registerAgent(new ResearcherAgent(llm, searchTool));
supervisor.registerAgent(new CoderAgent(llm));
supervisor.registerAgent(new ReviewerAgent(llm));

async function supervisorNode(state: AgentState): Promise<Partial<AgentState>> {
  const taskType = supervisor.analyzeTaskType(state.task);

  if (taskType === "code_development") {
    return { nextNode: "researcher", taskType };
  } else if (taskType === "research") {
    return { nextNode: "researcher", taskType };
  } else {
    return { nextNode: "researcher", taskType };
  }
}

async function researcherNode(state: AgentState): Promise<Partial<AgentState>> {
  if (!supervisor.agents.has("Researcher")) {
    return {};
  }

  const researcher = supervisor.agents.get("Researcher")!;
  const task = state.taskType === "code_development"
    ? `研究如何${state.task}`
    : state.task;

  const researchMessage = researcher.sendMessage("Supervisor", task);
  const researchResponse = await researcher.receiveMessage(researchMessage);

  if (researchResponse) {
    return {
      researchReport: researchResponse.content,
      messages: [
        ...state.messages,
        { role: "assistant", content: `[Researcher] ${researchResponse.content.slice(0, 200)}...` }
      ]
    };
  }

  return {};
}

async function coderNode(state: AgentState): Promise<Partial<AgentState>> {
  if (!supervisor.agents.has("Coder")) {
    return {};
  }

  const coder = supervisor.agents.get("Coder")!;
  const codeTask = state.researchReport
    ? `根据以下研究报告编写代码：\n\n${state.researchReport}\n\n任务：${state.task}`
    : state.task;

  const codeMessage = coder.sendMessage("Supervisor", codeTask);
  const codeResponse = await coder.receiveMessage(codeMessage);

  if (codeResponse) {
    return {
      codeContent: codeResponse.content,
      messages: [
        ...state.messages,
        { role: "assistant", content: `[Coder] ${codeResponse.content.slice(0, 200)}...` }
      ]
    };
  }

  return {};
}

async function reviewerNode(state: AgentState): Promise<Partial<AgentState>> {
  if (!supervisor.agents.has("Reviewer") || !state.codeContent) {
    return {};
  }

  const reviewer = supervisor.agents.get("Reviewer")!;
  const reviewMessage = reviewer.sendMessage("Supervisor", state.codeContent);
  const reviewResponse = await reviewer.receiveMessage(reviewMessage);

  if (reviewResponse) {
    return {
      reviewReport: reviewResponse.content,
      messages: [
        ...state.messages,
        { role: "assistant", content: `[Reviewer] ${reviewResponse.content.slice(0, 200)}...` }
      ]
    };
  }

  return {};
}

async function summaryNode(state: AgentState): Promise<Partial<AgentState>> {
  const results = [
    state.researchReport,
    state.codeContent,
    state.reviewReport
  ].filter(Boolean);

  const summaryPrompt = `作为 Supervisor，请汇总以下任务执行结果：

用户任务：${state.task}

执行结果：
${results.map((r, i) => `${i + 1}. ${r?.slice(0, 300)}...`).join("\n")}

请提供：
1. 任务完成情况
2. 关键成果
3. 建议
4. 下一步行动`;

  const response = await llm.invoke(summaryPrompt);

  return {
    messages: [
      ...state.messages,
      { role: "assistant", content: response.content as string }
    ]
  };
}

function shouldGoToCoder(state: AgentState): string {
  return state.taskType === "code_development" ? "coder" : "summary";
}

function shouldGoToReviewer(state: AgentState): string {
  return state.codeContent ? "reviewer" : "summary";
}

const workflow = new StateGraph<AgentState>({
  channels: {
    messages: {
      value: (x: AgentState["messages"], y?: AgentState["messages"]) => y ?? x,
      default: () => []
    },
    task: {
      value: (x: string, y?: string) => y ?? x,
      default: () => ""
    },
    researchReport: {
      value: (x: string | undefined, y?: string | undefined) => y ?? x ?? "",
      default: () => ""
    },
    codeContent: {
      value: (x: string | undefined, y?: string | undefined) => y ?? x ?? "",
      default: () => ""
    },
    reviewReport: {
      value: (x: string | undefined, y?: string | undefined) => y ?? x ?? "",
      default: () => ""
    },
    taskType: {
      value: (x: string | undefined, y?: string | undefined) => y ?? x ?? "",
      default: () => ""
    },
    nextNode: {
      value: (x: string | undefined, y?: string | undefined) => y ?? x ?? "",
      default: () => ""
    }
  }
})
  .addNode("supervisor", supervisorNode)
  .addNode("researcher", researcherNode)
  .addNode("coder", coderNode)
  .addNode("reviewer", reviewerNode)
  .addNode("summary", summaryNode)
  .addEdge("__start__", "supervisor")
  .addEdge("supervisor", "researcher")
  .addConditionalEdges("researcher", shouldGoToCoder, {
    coder: "coder",
    summary: "summary"
  })
  .addConditionalEdges("coder", shouldGoToReviewer, {
    reviewer: "reviewer",
    summary: "summary"
  })
  .addEdge("reviewer", "summary")
  .addEdge("summary", END);

export const app = workflow.compile();

File: langchain-python/09-multi-agent/graph.py
"""LangGraph 工作流定义"""

import os
import sys
from pathlib import Path
from typing_extensions import TypedDict
from typing import Annotated, Sequence, Literal
from dotenv import load_dotenv
from pydantic import SecretStr
from langgraph.graph import StateGraph, END
import operator

sys.path.insert(0, str(Path(__file__).parent))

from clients import create_model_client, create_search_tool
from agents import SupervisorAgent, ResearcherAgent, CoderAgent, ReviewerAgent

load_dotenv(override=True)


class AgentState(TypedDict):
    """智能体状态"""
    messages: Annotated[Sequence, operator.add]
    task: str
    research_report: str
    code_content: str
    review_report: str
    task_type: str
    next_node: str


def get_llm():
    """获取 LLM 实例"""
    openai_api_key = os.getenv("OPENAI_API_KEY", "")
    openai_base_url = os.getenv("OPENAI_BASE_URL", "https://api.openai.com/v1")
    model_name = os.getenv("MODEL_NAME", "gpt-3.5-turbo")
    
    from langchain_openai import ChatOpenAI
    
    return ChatOpenAI(
        model=model_name,
        temperature=0,
        api_key=SecretStr(openai_api_key),
        base_url=openai_base_url,
    )


def get_supervisor():
    """获取 Supervisor 实例"""
    llm = get_llm()
    search_tool = create_search_tool()
    
    supervisor = SupervisorAgent(llm)
    supervisor.register_agent(ResearcherAgent(llm, search_tool))
    supervisor.register_agent(CoderAgent(llm))
    supervisor.register_agent(ReviewerAgent(llm))
    
    return supervisor


async def supervisor_node(state: AgentState) -> dict:
    """Supervisor 节点"""
    supervisor = get_supervisor()
    
    task = state.get("task", "")
    if not task and state.get("messages"):
        last_message = state["messages"][-1]
        if isinstance(last_message, dict):
            task = last_message.get("content", "")
    
    task_type = supervisor.analyze_task_type(task)
    
    if task_type == "code_development":
        return {"next_node": "researcher", "task_type": task_type, "task": task}
    elif task_type == "research":
        return {"next_node": "researcher", "task_type": task_type, "task": task}
    else:
        return {"next_node": "researcher", "task_type": task_type, "task": task}


async def researcher_node(state: AgentState) -> dict:
    """Researcher 节点"""
    supervisor = get_supervisor()
    
    if "Researcher" not in supervisor.agents:
        return {}
    
    researcher = supervisor.agents["Researcher"]
    task = state.get("task", "")
    if not task and state.get("messages"):
        for msg in reversed(state["messages"]):
            if isinstance(msg, dict) and msg.get("content"):
                task = msg["content"]
                break
    
    if state.get("task_type") == "research":
        research_task = task
    else:
        research_task = f"研究如何{task}"
    
    research_message = researcher.send_message("Supervisor", research_task)
    research_response = await researcher.receive_message(research_message)
    
    if research_response:
        return {
            "research_report": research_response.content,
            "messages": [
                *state["messages"],
                {"role": "assistant", "content": f"[Researcher] {research_response.content[:200]}..."}
            ]
        }
    
    return {}


async def coder_node(state: AgentState) -> dict:
    """Coder 节点"""
    supervisor = get_supervisor()
    
    if "Coder" not in supervisor.agents:
        return {}
    
    coder = supervisor.agents["Coder"]
    task = state.get("task", "")
    research_report = state.get("research_report", "")
    
    if research_report:
        code_task = f"根据以下研究报告编写代码：\n\n{research_report}\n\n任务：{task}"
    else:
        code_task = task
    
    code_message = coder.send_message("Supervisor", code_task)
    code_response = await coder.receive_message(code_message)
    
    if code_response:
        return {
            "code_content": code_response.content,
            "messages": [
                *state["messages"],
                {"role": "assistant", "content": f"[Coder] {code_response.content[:200]}..."}
            ]
        }
    
    return {}


async def reviewer_node(state: AgentState) -> dict:
    """Reviewer 节点"""
    supervisor = get_supervisor()
    
    if "Reviewer" not in supervisor.agents or not state.get("code_content"):
        return {}
    
    reviewer = supervisor.agents["Reviewer"]
    review_message = reviewer.send_message("Supervisor", state["code_content"])
    review_response = await reviewer.receive_message(review_message)
    
    if review_response:
        return {
            "review_report": review_response.content,
            "messages": [
                *state["messages"],
                {"role": "assistant", "content": f"[Reviewer] {review_response.content[:200]}..."}
            ]
        }
    
    return {}


async def summary_node(state: AgentState) -> dict:
    """汇总节点"""
    results = [
        state.get("research_report"),
        state.get("code_content"),
        state.get("review_report")
    ]
    results = [r for r in results if r]
    
    task = state.get("task", "")
    if not task and state.get("messages"):
        for msg in reversed(state["messages"]):
            if isinstance(msg, dict) and msg.get("content"):
                task = msg["content"]
                break
    
    summary_prompt = f"""作为 Supervisor，请汇总以下任务执行结果：

用户任务：{task}

执行结果：
{chr(10).join([f"{i+1}. {r[:300]}..." for i, r in enumerate(results)])}

请提供：
1. 任务完成情况
2. 关键成果
3. 建议
4. 下一步行动"""
    
    llm = get_llm()
    response = await llm.ainvoke(summary_prompt)
    
    return {
        "messages": [
            *state["messages"],
            {"role": "assistant", "content": response.content}
        ]
    }


def should_go_to_coder(state: AgentState) -> Literal["coder", "summary"]:
    """判断是否转到 Coder"""
    return "coder" if state["task_type"] == "code_development" else "summary"


def should_go_to_reviewer(state: AgentState) -> Literal["reviewer", "summary"]:
    """判断是否转到 Reviewer"""
    return "reviewer" if state.get("code_content") else "summary"


def build_graph():
    """构建工作流图"""
    workflow = StateGraph(AgentState)
    
    workflow.add_node("supervisor", supervisor_node)
    workflow.add_node("researcher", researcher_node)
    workflow.add_node("coder", coder_node)
    workflow.add_node("reviewer", reviewer_node)
    workflow.add_node("summary", summary_node)
    
    workflow.set_entry_point("supervisor")
    
    workflow.add_edge("supervisor", "researcher")
    
    workflow.add_conditional_edges(
        "researcher",
        should_go_to_coder,
        {
            "coder": "coder",
            "summary": "summary"
        }
    )
    
    workflow.add_conditional_edges(
        "coder",
        should_go_to_reviewer,
        {
            "reviewer": "reviewer",
            "summary": "summary"
        }
    )
    
    workflow.add_edge("reviewer", "summary")
    workflow.add_edge("summary", END)
    
    app = workflow.compile()
    
    return app


app = build_graph()
